Map transaction status to different HTTP response codes

andrenpaes opened the pull request:
Also, centralize exception handling for plugin calls thru `PaymentPluginDispatcher.dispatchWithExceptionHandling()`, so TimeOut exceptions happening during payment/control plugin calls are translated into a `ErrorCode.PAYMENT_PLUGIN_TIMEOUT` exception.

That simplified exception handling in general, but it has a few subtleties:
* `OperationControlCallback`'s dispatched task is expected to always throw an `OperationException` with the [correct result](https://github.com/killbill/killbill/pull/534/files#diff-f8373089685a563920750c610e88fbc3R148) (FAILURE, EXCEPTION, etc)
* An `OperationException` thrown by the dispatched task is wrapped in a `PaymentApiException` and unwrapped by [`unwrapExceptionFromDispatchedTask`](https://github.com/killbill/killbill/pull/534/files#diff-f8373089685a563920750c610e88fbc3R156)
* Time outs, interrupted and lock failed exceptions are wrapped in `PaymentApiException` and don't trigger a Plugin `onFailure` call (the part about not calling `onFailure` hasn't changed, it was already the behavior of the code)

pierre commented on jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/JaxRsResourceBase.java:
```diff
@@ -533,4 +535,48 @@ protected void verifyNumberOfElements(int actual, int expected, String errorMess
+                return uriBuilder.buildResponse(uriInfo, PaymentResource.class, "getPayment", payment.getId());
+            case PAYMENT_FAILURE:
+                // 402 - Payment Required
+                return Response.status(402).build();
```
Could we keep the location header in failure scenarios too? That's the only way for the client to get the payment id (if no external key was specified).

sbrossie replied:
I am actually not convinced about that one because the client has the ability to pass an `externalKey` which can be used to retrieve the payment. In theory `Location` header are specific to `201`, so unless we have good reason to twist http spec, i'd be more inclined to not return it (but i don't have a strong feeling, as long as we are consistent across our apis, which means that if we decide to return it, we should do a pass across all our apis to make sure this matches everywhere).

pierre replied:
One main use-case is to keep the `refresh` functionality in the clients. Even if we now return different return codes for the various states, the clients will still most likely want to access the payment object (to access the gateway error message for example).

sbrossie replied:
I think we should talk about that and make sure we agree on what is the behavior we want. Probably, this is good enough for that upcoming release, and then we can do another pass on that topic.

pierre commented on jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/JaxRsResourceBase.java:
```diff
@@ -533,4 +535,48 @@ protected void verifyNumberOfElements(int actual, int expected, String errorMess
+
+        if (transactionExternalKey != null) {
+            for (final PaymentTransaction transaction : matchingTransactions) {
+                if (Objects.equal(transaction.getExternalKey(), transactionExternalKey)) {
```
Sure, that works, but since we know `transactionExternalKey != null`, you could simply have `transactionExternalKey.equals(transaction.getExternalKey())`.

pierre commented on jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/JaxRsResourceBase.java:
```diff
@@ -533,4 +535,48 @@ protected void verifyNumberOfElements(int actual, int expected, String errorMess
 
+    protected Response createPaymentResponse(final UriInfo uriInfo, final Payment payment, final TransactionType transactionType, @Nullable final String transactionExternalKey) {
+        final PaymentTransaction createdTransaction = findCreatedTransaction(payment, transactionType, transactionExternalKey);
+        Preconditions.checkNotNull(createdTransaction, "No transaction of type '%s' found", transactionType);
```
This will probably return 5xx. Shouldn't we return 404 instead?

sbrossie replied:
That's actually a tricky one: Does it fail because client passed the wrong `transactionType` (=> `404`) or because our state is inconsistent (=> `5xx`). It would make sense to assume the former, but this is so not obvious.

Also it would be nice to use some kind of Preconditions in which case we might need to tweak the mapping into http status?

andrenpaes replied:
I don't expect this to ever be triggered, since we're always executing this code after a payment transaction was created, so a transaction should have been created, or at least already existed.

So, if this is triggered, it means something is wrong elsewhere in kb, unless I'm missing something.

pierre replied:
> So, if this is triggered, it means something is wrong elsewhere in kb, unless I'm missing something.

What about the aborted case? Wouldn't it trigger a NPE in `findCreatedTransaction`?

andrenpaes replied:
Yeah, that is still broken, but now is broken uniformly :)

We could return a No Content response, but I'm not sure how helpful would that be without some hint on why the payment was aborted. Also, it's impossible to tell if the payment is null b/c it was aborted, or b/c some regression was introduced into code.

I believe throwing a NPE is probably the best (or least worst) way to go, since we don't ever expect aborted payments right now, and if it happens it will be easy it to notice.

pierre replied:
> we don't ever expect aborted payments right now

Well, who is `we`? 😸  It's up to plugin developers to abort payments in their control plugin if needed (e.g. [fraud detection systems](https://github.com/killbill/killbill-accertify-plugin/blob/master/src/main/java/org/killbill/billing/plugin/accertify/api/AccertifyPaymentControlPluginApi.java#L71-L81)).

andrenpaes replied:
I understand that, and I'm not saying we shouldn't fix. But I think we need a better solution to differentiate a `null` return from an aborted payment.
I think we'd benefit from having a separate issue to cover this as this PR is introducing a lot of changes already.

andrenpaes replied:
Just to close the loop here, I'll be working on handling the abort payment scenario in another PR.

pierre commented on payment/src/test/java/org/killbill/billing/payment/core/sm/TestPluginOperation.java:
```diff
@@ -57,7 +57,7 @@
     private static final String PLUGIN_NAME_PLACEHOLDER = "pluginName";
 
-    private static final int TIMEOUT = 5;
+    private static final int TIMEOUT = 10;
```
Why this change?

andrenpaes replied:
I was getting timeout exceptions while running some of the tests of this class.

pierre commented on payment/src/main/java/org/killbill/billing/payment/core/sm/control/OperationControlCallback.java:
```diff
@@ -149,33 +145,19 @@ public OperationResult doOperationCallback() throws OperationException {
-            logger.warn("Call was interrupted for accountId='{}'", paymentStateContext.getAccount().getId());
-        } else {
-            logger.warn("Operation failed for accountId='{}'", paymentStateContext.getAccount().getId(), e);
+    protected OperationException unwrapExceptionFromDispatchedTask(final PaymentApiException e) {
```
Do we still need these `unwrapExceptionFromDispatchedTask`? There is no unwrapping done at this level anymore (at least the name should be changed).

andrenpaes replied:
We still need to either unwrap the `PaymentApiException` returned or wrap it into an `OperationException` instance.
Initially I've considered packing everything in `OperationCallbackBase` class, but the `PaymentOperation` class does some things of its own with errors thrown, so I'm not sure we can get rid of it.

andrenpaes replied:
This part can be further simplified too, but I'll open a different PR so it's easier to test and review.

pierre commented on payment/src/main/java/org/killbill/billing/payment/core/sm/payments/PaymentOperation.java:
```diff
@@ -86,35 +82,16 @@ public OperationResult doOperationCallback() throws OperationException {
-                throw unwrapExceptionFromDispatchedTask(e);
-            }
-        }
+	        } catch (final OperationException e) {
```
Alignment issues?

andrenpaes replied:
Tabs! :)
I'll fix it.

pierre replied:
![the-simpsons-episode-25 02-treehouse-of-horror-xxiv-promotional-photos-1_full](https://cloud.githubusercontent.com/assets/51940/15024500/fa2d898e-11e9-11e6-8b43-3bce5adcaab1.jpg)

pierre commented:
After re-reading the code and the diff, I don't see any major red flag (and 👍 if we can simplify our code). However, having recently spent a good amount of time in this piece of code, I also know it's quite easy to overlook a code path (we have a matrix of combinations, with the two state machines and the list of exceptions that can be thrown at various points)...

If it hasn't been done, we should at least do a round of manual testing. When I did my recent logging changes, I triggered manually:

* A state machine timeout with a control plugin taking too much time (before the request goes to the payment plugin)
* Ruby and Java Payment plugins timeouts
* Ruby and Java Payment plugins throwing a runtime exception

Each time, I verified the logs were meaningful (no unnecessary wrapping of exceptions or unnecessary logging of java.util exceptions -- see https://github.com/killbill/killbill/issues/394 for example).

Additionally:

* At the Java API level, what is the current status of our tests for this exception handling? Are we confident we have enough?
* Do the server tests pass as-is without any client changes? Can we add some tests when the mock plugin timeouts, throws an exception, etc. if we don't have any?

sbrossie commented on jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/JaxRsResourceBase.java:
```diff
@@ -533,4 +535,48 @@ protected void verifyNumberOfElements(int actual, int expected, String errorMess
+        // If nothing is found, return the latest transaction of given type
+        return Iterables.getFirst(matchingTransactions, null);
+    }
 }
```
Nit: Also as i mentioned yesterday on previous PR, this could be simplified into one iteration (instead of 3).

andrenpaes replied:
We can definitely optimize it to do it in 1 iteration, but there's no performance gain in doing it.

My priority here was to communicate as clear as possible what the code is doing.

andrenpaes commented on profiles/killbill/src/test/resources/killbill.properties:
```diff
@@ -21,6 +21,8 @@ org.killbill.overdue.uri=overdue.xml
 
 org.killbill.payment.retry.days=8,8,8
 
+org.killbill.payment.plugin.timeout=8s
```
The default timeout of the java client is 10 seconds, so I've lowered the threshold in kb (from the default 30s), so we could add a test case for timeouts.

sbrossie replied:
I think it would be better to see if we could overwrite the timeout value for that test specifically to something even smaller:

* `8 sec` is still quite long for a test that runs every time
* Also, that impacts all out tests (where default timeout is now 8 sec)

andrenpaes replied:
I think this only impact the tests in the profile module. As far as I could tell, all tests that use this setting are in the slow group.

But yeah, it would be great to scope this to the test. I didn't want to set it lower since I was afraid we'd start getting timeouts in other tests that don't expect timeouts.

andrenpaes commented:
Ok, so I've added test cases covering most of the error handling scenarios. I've also cleaned up the logging and fixed a small issue in the `PaymentOperation`/`PaymentAutomaton` classes.

Regarding the responses, I've added the location header and error messages to the response payload. The java client is now able to handle the new error response correctly (by correctly, I mean raising a `KillBillClientException`).

Finally, I'll leave the abort payment handling to a different PR, since the changes here won't influence that case, as the latest changes for 0.16.4 introduced a regression in the `DefaultPaymentApi` class:
https://github.com/killbill/killbill/blob/master/payment/src/main/java/org/killbill/billing/payment/api/DefaultPaymentApi.java#L128-L139

I think this is ready to be merged. Please take a look and let me know if you find any issues.

sbrossie commented:
@andrenpaes Looks good t me. 👍  on the new tests. Just have one comment about that default timeout value, but that's not a blocker, we can talk about it. If @pierre agrees, i am ready to merge

sbrossie commented:
This had been merged in branch `work-for-release-0.16.5`. Closing PR.
